### ISSUE TITLE
Use type table to index value types

### ISSUE THREAD

**Issue by AndrewScheidecker** [WebAssembly/design#749]
The roadmap for WASM anticipates that value types will eventually include composite types (function pointer types are one example). However, value types are currently encoded as a 1 byte value, which is inadequate for expressing arbitrary composite types.

Additionally, defining a static enumeration of value types in the standard means that the encoding needs to allow for the universe of possible value types, not just those used by the module.

I propose that a value types should be encoded as a varuint index into the type table. In addition to the current function type constructors in the type table, there should be a basic type constructor that takes a string argument. The MVP would only define basic types for "i32", "i64", "f32", and "f64", but could eventually support things like "f16x32", "i8", etc.

**Comment by rossberg-chromium** [comment]
In general, we will have a grammar for types, and their binary representation will be a (serialised) AST of sorts, whose "opcodes" are the different type constructors (and not all of them have 0 operands!). We currently just have a very rudimentary and fragmented grammar:

- 4 constructors for value types
- 1 constructor for table element types (`anyfunc`)
- 1 constructor for type definitions (function types)

Currently, these 3 classes are separated, but we should not assume them to stay that way. They are e.g. likely to mix in the definition of field types for structs, along with other new constructors. That's why we have made sure that their binary codes are disjoint, even though that is currently unnecessary.

What precise grammar we are going to get in the future is of course still open, but the current encoding generalises cleanly. E.g., one likely addition will be to allow values to have a reference type `ref <id>`, where `<id>` is an index into the type section. In fact, there may potentially be different reference types that can be formed for the same struct def (flat or interior ptr) or function sig (ptr or closure).

In other words, there will eventually be a (larger) grammar for value types, and it includes references to named types. But it will also include other things.

**Comment by AndrewScheidecker** [comment]
The problem with extending the existing value type serialization to include more constructors is that you need to allocate tags for those constructors globally. If value types are stored in the type table, there's a simple and compact encoding for each use of the value type, and the entry in the type table can use a string. And IMO strings are a bit less scary to assign global meaning to than some finite index space.

**Comment by rossberg-chromium** [comment]
In terms of cardinality, the index space of strings is no larger than that of varuint. ;) However, it is more expensive to decode. What is different about types compared to, say, expressions that would justify a special encoding? In fact, we just reverted the decision to use strings instead of numbers to identify sections.

**Comment by pizlonator** [comment]
It's too late for this.
